payroll/models/hr_payslip.py

```diff
@@ -0,0 +1,581 @@
+            raise UserError(
+                _("You cannot delete a payslip which is not draft or " "cancelled!")
+            )
+        return super(HrPayslip, self).unlink()
+
+    # TODO move this function into hr_contract module, on hr.employee object
+    @api.model
+    def get_contract(self, employee, date_from, date_to):
+        """
+        @param employee: recordset of employee
+        @param date_from: date field
+        @param date_to: date field
+        @return: returns the ids of all the contracts for the given employee
+                 that need to be considered for the given dates
+        """
+        # a contract is valid if it ends between the given dates
+        clause_1 = ["&", ("date_end", "<=", date_to), ("date_end", ">=", date_from)]
+        # OR if it starts between the given dates
+        clause_2 = ["&", ("date_start", "<=", date_to), ("date_start", ">=", date_from)]
+        # OR if it starts before the date_from and finish after the date_end
+        # (or never finish)
+        clause_3 = [
+            "&",
+            ("date_start", "<=", date_from),
+            "|",
+            ("date_end", "=", False),
+            ("date_end", ">=", date_to),
+        ]
+        clause_final = (
+            [("employee_id", "=", employee.id), ("state", "=", "open"), "|", "|"]
+            + clause_1
+            + clause_2
+            + clause_3
+        )
+        return self.env["hr.contract"].search(clause_final).ids
+
+    def compute_sheet(self):
+        for payslip in self:
+            number = payslip.number or self.env["ir.sequence"].next_by_code(
+                "salary.slip"
+            )
+            # delete old payslip lines
+            payslip.line_ids.unlink()
+            # set the list of contract for which the rules have to be applied
+            # if we don't give the contract, then the rules to apply should be
+            # for all current contracts of the employee
+            contract_ids = payslip.contract_id.ids or self.get_contract(
+                payslip.employee_id, payslip.date_from, payslip.date_to
+            )
+            lines = [
+                (0, 0, line)
+                for line in self._get_payslip_lines(contract_ids, payslip.id)
+            ]
+            payslip.write({"line_ids": lines, "number": number})
+        return True
+
+    @api.model
+    def get_worked_day_lines(self, contracts, date_from, date_to):
+        """
+        @param contract: Browse record of contracts
+        @return: returns a list of dict containing the input that should be
+                 applied for the given contract between date_from and date_to
+        """
+        res = []
+        # fill only if the contract as a working schedule linked
+        for contract in contracts.filtered(
+            lambda contract: contract.resource_calendar_id
+        ):
+            day_from = datetime.combine(fields.Date.from_string(date_from), time.min)
```

Comment:
```suggestion
            day_from = datetime.combine(date_from, time.min)
```
version 12+ Odoo returns always native Python objects

ref : https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-12.0#tasks-to-do-in-the-migration